src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -2316,12 +2314,46 @@ void G1CollectedHeap::collection_set_iterate_all(HeapRegionClosure* cl) {
   _collection_set.iterate(cl);
 }
 
-void G1CollectedHeap::collection_set_par_iterate_all(HeapRegionClosure* cl, HeapRegionClaimer* hr_claimer, uint worker_id) {
-  _collection_set.par_iterate(cl, hr_claimer, worker_id, workers()->active_workers());
+void G1CollectedHeap::collection_set_par_iterate_all(HeapRegionClosure* cl,
+                                                     HeapRegionClaimer* hr_claimer,
+                                                     uint worker_id) {
+  _collection_set.par_iterate(cl, hr_claimer, worker_id);
 }
 
-void G1CollectedHeap::collection_set_iterate_increment_from(HeapRegionClosure *cl, HeapRegionClaimer* hr_claimer, uint worker_id) {
-  _collection_set.iterate_incremental_part_from(cl, hr_claimer, worker_id, workers()->active_workers());
+void G1CollectedHeap::collection_set_iterate_increment_from(HeapRegionClosure *cl,
+                                                            HeapRegionClaimer* hr_claimer,
+                                                            uint worker_id) {
+  _collection_set.iterate_incremental_part_from(cl, hr_claimer, worker_id);
+}
+
+void G1CollectedHeap::par_iterate_regions_array_part_from(HeapRegionClosure* cl,
```

Comment:
This method essentially apply a closure to all elements in `[offset, length)` of `regions`. I wonder if we can pass `&regions[offset]` and `length-offset` to this method. Then this method can work on the whole array, and figuring out the starting index for each worker becomes trivial, `worker_id * array_len / #workers`. It's not obvious to me why the calculation in this patch guarantees balanced start index.

Reply:
In fact, this method is just moved from G1CollectionSet::iterate_part_from which is deleted, the movement is to simply reuse its logic in both G1CollectionSet and G1EvacFailureRegions.

Your suggestion seems reasonable to me, thanks for catching this.  
As this is an refinement to existing code, I wonder if we could do this in another issue? And as I know Thomas is starting another refactoring in G1's related code, so I hope this can be pushed eariler to avoid further merge conflicting in either pr. 
How do you think about it? I'll create an issue to track this code refinement if you agree.

Reply:
> this method is just moved from

Yes, you are right; I didn't realize it.

> How do you think about it? I'll create an issue to track this code refinement if you agree.

Sure; thank you.